File: 279_Perfect_Squares/279_1.java
class Solution {
    public int numSquares(int n) {
        int len = (int)Math.sqrt(n);
        List<Integer> squares = new ArrayList<>();
        for (int i = 1; i <= len; i++) {
            squares.add(i * i);
        }

        Set<Integer> queue = new HashSet<>();
        queue.add(n);
        int level = 0;
        while (!queue.isEmpty()) {
            level++;
            Set<Integer> nextQueue = new HashSet<>();
            for (Integer num : queue) {
                for (Integer squ : squares) {
                    if (num.equals(squ)) {
                        return level;
                    } else if (num < squ) {
                        break;
                    } else {
                        nextQueue.add(num - squ);
                    }
                }
            }
            queue = nextQueue;
        }
        return n;
    }
}

File: 279_Perfect_Squares/279_1.py
class Solution:
    def numSquares(self, n: int) -> int:
        squares = [i**2 for i in range(1, int(sqrt(n)) + 1)]
        dp = [float("inf")] * (n + 1)
        dp[0] = 0
        for i in range(1, n + 1):
            for squ in squares:
                if squ > i:
                    break
                dp[i] = min(dp[i], dp[i - squ] + 1)
        return dp[-1]
